### DISCUSSION TITLE
Updates to the codebase

### DISCUSSION THREAD

**Discussion by MendyBerger** [ji-devs/ji-cloud#1785] [Ideas]
These are thinks that might be helpful to update/add in the codebase.

We probably won't end up doing all of it, this is more like food for thought.

- Fix all compiler warnings.
- Integrate `cargo fmt`.
- Integrate `cargo clippy`.
- Prettier.
- Eslint.
- Remove circular dependencies.
- Change to Rust stable (I think that we're currently on nightly).
- Update to Rust 2021.
- Update to lit 3.
- New branch names: `prod`, `stage`, `dev` (don't have a real reason for those names besides being easier on **my** brain).
- Block dev and stage with "Basic access authentication" (https://docs.rs/actix-web-httpauth/ might help).
- Integrate Google Cloud Armor
- Can we minify the html and css of lit components?

**Comment by dakom** [comment]
Some more thoughts:

- code style guide and adherence to it (self: Rc<Self>, actions vs. state vs. dom, etc.)
- CI should run to completion for frontend and backend before deploying either, at least on `prod`

**Comment by MendyBerger** [comment]
About moving to stable, I think we need to wait for [type_alias_impl_trait](https://github.com/rust-lang/rust/issues/63063) to stabilize, we use this feature a lot 
```rust
impl Signal<Item = SomeType>;
```

**Comment by dakom** [comment]
I think that RFC is more about so-called "existential types" via the impl Trait semantics.

So with your example, it would be more about being able to name that signal like `type SomeSignal = impl Signal<Item = SomeType>` and then pass that `SomeSignal` into a function, or hold it in a struct, etc. Kinda like `type SomeSignal = Box<dyn Signal...>` but without the performance hit or having to satisfy the Pin requirements etc.

Here's a working playground example with futures: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=630055fe5b3e9b99bc22296ec5f3834e

Existential types have their own constraints, but that's a bit outside the scope here.

We were using it in the module abstraction thing, but ultimately I can't remember if it's still required.

**Comment by MendyBerger** [comment]
Have a file for common elements that can be shared between entries, would enable caching of those elements

**Reply by dakom** [reply]
That was actually the original intent of _bundles dir iirc, not just as a convenience for importing, but for actually exporting the bundles as separate files
